ndcube/analysis/rebin.py

```diff
@@ -0,0 +1,413 @@
+        tmp_uncertainties = []
+        tmp_coords = [[]] * len(new_coord_values)
+        item = [slice(None)] * ndim
+        coord_item = np.array([slice(None)] * ndim)
+        for j in range(len(edges) - 1):
+            # Rebin data
+            item[i] = slice(edges[j], edges[j + 1])
+            tuple_item = tuple(item)
+            data_chunk = new_data[tuple_item]
+            tmp_data.append(operation(data_chunk, axis=i))
+            # Rebin mask
+            if rebin_mask is True:
+                mask_chunk = new_mask[tuple_item]
+                tmp_mask.append(handle_mask(mask_chunk, axis=i))
+            # Rebin uncertainties
+            if propagate_uncertainties:
+                # Reorder axis i to be 0th axis as this is format required by propagate_rebin_uncertainties
+                uncertainty_chunk = type(new_uncertainty)(
+                    np.moveaxis(new_uncertainty.array[tuple_item], i, 0), unit=new_uncertainty.unit
+                )
+                data_unc = np.moveaxis(data_chunk, i, 0)
+                mask_unc = np.moveaxis(mask_chunk, i, 0) if rebin_mask else new_mask
+                tmp_uncertainty = utils.cube.propagate_rebin_uncertainties(
+                    uncertainty_chunk,
+                    data_unc,
+                    mask_unc,
+                    operation,
+                    operation_ignores_mask=operation_ignores_mask,
+                    handle_mask=handle_mask,
+                    new_unit=new_unit,
+                    **kwargs,
+                )
+                tmp_uncertainties.append(tmp_uncertainty.array)
+            # Rebin coordinates
+            coord_item[i] = np.array([edges[j], edges[j + 1] - 1])
+            for k, (coord, coord_arr_idx) in enumerate(zip(new_coord_values, wc_arr_idxs)):
+                if i in coord_arr_idx:
+                    idx_i = np.where(coord_arr_idx == i)[0][0]
+                    tmp_coords[k].append(np.mean(coord[tuple(coord_item[coord_arr_idx])], axis=idx_i))
+        # Restore rebinned axes.
+        if rebin_mask:
+            new_mask = np.stack(tmp_mask, axis=i)
+        new_data = np.ma.masked_array(np.stack(tmp_data, axis=i).data, mask=new_mask)
+        if propagate_uncertainties:
+            new_uncertainty = type(new_uncertainty)(np.stack(tmp_uncertainties, axis=i), unit=new_uncertainty.unit)
+        for k, coord_arr_idx in enumerate(wc_arr_idxs):
+            if i in coord_arr_idx:
+                idx_i = np.where(coord_arr_idx == i)[0][0]
+                new_coord_values[k] = np.stack(tmp_coords[k], axis=idx_i)
+
+    # Build new WCS.
+    ec = ExtraCoords()
+    for name, axis, coord, physical_type in zip(
+        new_coord_names, new_coord_pix_idxs, new_coord_values, new_coord_physical_types
+    ):
+        if len(axis) != 1:
+            raise NotImplementedError("Rebinning multi-dimensional coordinates not supported.")
+        ec.add(name, axis, coord, physical_types=physical_type)
+    new_wcs = ec.wcs
```

Comment:
Using extra coords as a way to do this is a bit "weird" using table coordinate directly might be more applicable.

Reply:
Would this require extra code to parse whether it should be a `QuantityTableCoordinate`, `TimeTableCoordinate`, etc.?  If so, `ExtraCoords` already does this for us.